testing/python/fixtures.py

```diff
@@ -3653,22 +3653,72 @@ def test_foo(request):
             )
         )
 
-        tests_dir.chdir()
         testdir.syspathinsert(fixdir)
+
+        # From root.
+        result = testdir.runpytest()
+        result.stdout.fnmatch_lines(
+            [
+                "The requested fixture has no parameter defined for test:",
+                "    tests/test_foos.py::test_foo",
+                "",
+                "Requested fixture 'fix_with_param' defined in:",
+                "fixtures/fix.py:4",
+                "Requested here:",
+                "tests/test_foos.py:4",
+                "*1 failed*",
+            ]
+        )
+
+        # With changed dir.
+        tests_dir.chdir()
         result = testdir.runpytest()
         result.stdout.fnmatch_lines(
             [
                 "The requested fixture has no parameter defined for test:",
                 "    test_foos.py::test_foo",
                 "",
                 "Requested fixture 'fix_with_param' defined in:",
-                "*fix.py:4",
+                "{}:4".format(fixfile),
                 "Requested here:",
                 "test_foos.py:4",
                 "*1 failed*",
             ]
         )
 
+        # With non-overlapping rootdir, passing tests_dir.
+        rootdir = testdir.mkdir("rootdir")
+        rootdir.chdir()
+        result = testdir.runpytest("--rootdir", rootdir, tests_dir)
+        result.stdout.fnmatch_lines(
+            [
+                "The requested fixture has no parameter defined for test:",
+                "    test_foos.py::test_foo",
+                "",
+                "Requested fixture 'fix_with_param' defined in:",
+                "{}:4".format(fixfile),
+                "Requested here:",
+                "{}:4".format(testfile),
+                "*1 failed*",
+            ]
+        )
+
+        # With non-overlapping rootdir, passing testfile.
+        result = testdir.runpytest("--rootdir", rootdir, testfile)
+        result.stdout.fnmatch_lines(
+            [
+                "The requested fixture has no parameter defined for test:",
+                # XXX: missing filename in funcitem.nodeid here!
```

Comment:
This appears to be a generic issue with nodeid handling it seems - will pull this out here.